fuse/readonly/readonly_unix.go

```diff
@@ -133,31 +143,84 @@ func (s *Node) Attr(ctx context.Context, a *fuse.Attr) error {
 // Lookup performs a lookup under this node.
 func (s *Node) Lookup(ctx context.Context, name string) (fs.Node, error) {
 	log.Debugf("Lookup '%s'", name)
-	nodes, err := s.Ipfs.Resolver.ResolveLinks(ctx, s.Nd, []string{name})
-	if err != nil {
+	link, _, err := uio.ResolveUnixfsOnce(ctx, s.Ipfs.DAG, s.Nd, []string{name})
+	switch err {
+	case os.ErrNotExist, mdag.ErrLinkNotFound:
 		// todo: make this error more versatile.
 		return nil, fuse.ENOENT
+	default:
+		log.Errorf("fuse lookup %q: %s", name, err)
+		return nil, fuse.EIO
+	case nil:
+		// noop
 	}
 
-	pbnd, ok := nodes[len(nodes)-1].(*mdag.ProtoNode)
-	if !ok {
-		log.Error("fuse lookup got non-protobuf node")
-		return nil, fuse.ENOTSUP
+	nd, err := s.Ipfs.DAG.Get(ctx, link.Cid)
+	switch err {
+	case mdag.ErrNotFound:
+	default:
+		log.Errorf("fuse lookup %q: %s", name, err)
+		return nil, err
+	case nil:
+		// noop
 	}
 
-	return &Node{Ipfs: s.Ipfs, Nd: pbnd}, nil
+	return &Node{Ipfs: s.Ipfs, Nd: nd}, nil
 }
 
 // ReadDirAll reads the link structure as directory entries
 func (s *Node) ReadDirAll(ctx context.Context) ([]fuse.Dirent, error) {
 	log.Debug("Node ReadDir")
-	entries := make([]fuse.Dirent, len(s.Nd.Links()))
-	for i, link := range s.Nd.Links() {
-		n := link.Name
+	dir, err := uio.NewDirectoryFromNode(s.Ipfs.DAG, s.Nd)
+	if err != nil {
+		return nil, err
+	}
+
+	var entries []fuse.Dirent
+	err = dir.ForEachLink(ctx, func(lnk *format.Link) error {
+		n := lnk.Name
 		if len(n) == 0 {
-			n = link.Cid.String()
+			n = lnk.Cid.String()
+		}
+		// TODO: calling everything a DT_File here might cause issues. But it
+		// will be expensive to query each child. However most shells call an
+		// additional 'stat' on each item in a directory listing, its probably
+		// okay.
+		nd, err := s.Ipfs.DAG.Get(ctx, lnk.Cid)
+		if err != nil {
+			log.Warning("error fetching directory child node: ", err)
+		}
+
+		var t fuse.DirentType
+		switch nd := nd.(type) {
+		case *mdag.RawNode:
+			t = fuse.DT_File
+		case *mdag.ProtoNode:
+			var data ftpb.Data
+			if err := proto.Unmarshal(nd.Data(), &data); err != nil {
+				log.Warning("failed to unmarshal protonode data field:", err)
+			} else {
+				switch data.GetType() {
+				case ftpb.Data_Directory, ftpb.Data_HAMTShard:
+					t = fuse.DT_Dir
+				case ftpb.Data_File, ftpb.Data_Raw:
+					t = fuse.DT_File
+				case ftpb.Data_Symlink:
+					t = fuse.DT_Link
+				case ftpb.Data_Metadata:
+					log.Error("metadata object in fuse should contain its wrapped type")
+				default:
+					log.Error("unrecognized protonode data type: ", data.GetType())
+				}
+			}
+		default:
+			t = fuse.DT_Unknown
```

Comment:
Maybe put this up top (i.e., `t := fuse.DT_Unknown`). We have several cases where we just log errors but don't assign anything (I assume it will default to unknown but that's not immediately obvious).